### ISSUE TITLE
Calculate version automatically from the input data and selected error correction level

### ISSUE THREAD

**Issue by monospaced** [monospaced/angular-qrcode#6]
On reflection, doesn't make much sense to allow combos of data, version, and error correction level that just result in over-capacity errors.

Would be more useful to ensure a valid qrcode is always created by looking-up and applying the appropriate version based on the data length (in bits) and selected error-correction-level. Could then give data a max-length of 23648, and it would always generate a valid qrcode up to that limit.

**Comment by arkarkark** [comment]
I'd be really interested in this too. It'd be even nicer if you could also select the ECC level for the best results depending on the data size.
1. first choose the lowest version that will hold your data
2. then choose the best level that will still hold your data

**Comment by arkarkark** [comment]
This is so ugly I'm not even going to submit a pull request. but...
Here's one way that just brute forces the qr generator until it gets a valid qr code with no error

```javascript
              version = 1;
              levelsArray = 'H Q M L'.split(' ');
              levelsIndex = 0

              while (true) {
                qr = qrcode(version, levelsArray[levelsIndex]);
                qr.addData(data);
                try {
                  qr.make();
                  error = false;
                  console.log('success', data.length, version, levelsArray[levelsIndex]);
                  break
                } catch(e) {
                  error = e.message;
                  levelsIndex++;
                  if (levelsIndex >= levelsArray.length) {
                    version++;
                    if (version > 10) {
                      return;
                    }
                    levelsIndex = 0;
                  }
                }
              }
```

**Comment by arkarkark** [comment]
alternatively this lookup table can be useful in your own code. I've found I need to lookup on the `binary` value for some reason.

```javascript
  var qrLookup = [ // from https://developers.google.com/chart/infographics/docs/qr_codes
    {version: 1, level: 'H', digits: 17, chars: 10, binary: 7, kanji: 4},
    {version: 1, level: 'Q', digits: 27, chars: 16, binary: 11, kanji: 7},
    {version: 1, level: 'M', digits: 34, chars: 20, binary: 14, kanji: 8},
    {version: 1, level: 'L', digits: 41, chars: 25, binary: 17, kanji: 10},
    {version: 2, level: 'H', digits: 34, chars: 20, binary: 14, kanji: 8},
    {version: 2, level: 'Q', digits: 48, chars: 29, binary: 20, kanji: 12},
    {version: 2, level: 'M', digits: 63, chars: 38, binary: 26, kanji: 16},
    {version: 2, level: 'L', digits: 77, chars: 47, binary: 32, kanji: 20},
    {version: 3, level: 'H', digits: 58, chars: 35, binary: 24, kanji: 15},
    {version: 3, level: 'Q', digits: 77, chars: 47, binary: 32, kanji: 20},
    {version: 3, level: 'M', digits: 101, chars: 61, binary: 42, kanji: 26},
    {version: 3, level: 'L', digits: 127, chars: 77, binary: 53, kanji: 32},
    {version: 4, level: 'H', digits: 82, chars: 50, binary: 34, kanji: 21},
    {version: 4, level: 'Q', digits: 111, chars: 67, binary: 46, kanji: 28},
    {version: 4, level: 'M', digits: 149, chars: 90, binary: 62, kanji: 38},
    {version: 4, level: 'L', digits: 187, chars: 114, binary: 78, kanji: 48}
  ];
```

```javascript
_.find(qrLookup, (q) -> q.binary  > data.length)
```

is useful.

**Comment by monospaced** [comment]
Hi @arkarkark thanks for the contribution, that is super useful :+1: 

Unfortunately I'm up to my ears in client work at the moment, so don't have a lot of time for open source right now :(

Hopefully in about a month or so I will get back into this, and look at implement the lookup solution :)

**Comment by monospaced** [comment]
Resurrecting this thread to comment that, thanks to @bgok (and also @arkarkark who essentially has the same idea) , fallback 'auto-minimum-version' functionality has now been added to master: 6f7abbeb03f9d

I've also added a warning when the version is incremented: dac73ed1df94f5dad697f1fb92870b72d8060a78

If at all possible, a sufficient `version` for the anticipated data should still be specified, to avoid unnecessarily triggering this brute force approach.

However, in the case where you simply don't know the amount of data that could be supplied (and also where you desire the lowest possible version to accommodate any data given) this does the job.

It doesn't tackle the Error Correction Level, which still needs to be understood and handled manually.

Definitely room for improvement, but on balance I do prefer this approach to simply failing when the version is too low.
